### PR TITLE
Handle invalid drv and other items (Apps-2039)

### PR DESCRIPTION
**Pull request by Fabtron**
### What's new?
If a deposit return voucher is invalid we need to handle it. Since we can't define that we received an invalid drv and only do a match up via the id from the error the message is adjusted to match that of a product.

We dialog is integrated in several places. E.g. in the selfscanning view to ensure that it will be removed directly by the user after it was added and in the checkoutbar to ensure that a checkout can't be started if a invalid product is part of the cart (this was the previous way). 

APPS-2039

### How to test?
Integrate this branch into an application and scan the barcode mentioned in the ticket.

### Definition of Done

- [x] Issue is linked
- [x] All requirements of the issue are fulfilled
- [x] Changelog is updated
- [x] Documentation is updated
- [x] [Self-Review](https://www.nerdwallet.com/blog/engineering/why-you-should-be-doing-self-reviews/)
- [ ] Review with the Product Owner _(Release-Variante o. Minified Build)_

#### App Tests
- [ ] Minified build has been tested _(aka. Release Build)_
- [ ] Environments have been taken care of _(Production/Staging)_
- [x] Supported languages have been tested
- [x] Light-/Dark-Mode has been tested
- [ ] Edge-Cases have been tested
- [ ] Android API Levels have been taken care of _(minSdk?)_

#### Testing
- [ ] Tests have been written _(aka Unit-Tests, Integration-Tests, ...)_

### CODE REVIEW

File: core/src/main/java/io/snabble/sdk/shoppingcart/ShoppingCart.kt [hunk 1]
```diff
@@ -442,6 +450,15 @@ class ShoppingCart(
...
         }
 
+    /**
+     * Gets a list of invalid item ids that were rejected by the backend.
+     */
+    var invalidItemIds: List<String>?
+        get() = data.invalidItemIds ?: emptyList()
+        set(invalidItemIds) {
+            data = data.copy(invalidItemIds = invalidItemIds)
+        }
```
Comment (cmaier) [inline core/src/main/java/io/snabble/sdk/shoppingcart/ShoppingCart.kt:460]: Cannot be null!?

File: core/src/main/java/io/snabble/sdk/shoppingcart/ShoppingCartUpdater.kt [hunk 2]
```diff
@@ -94,6 +100,14 @@ internal class ShoppingCartUpdater(
                     error(requestSucceeded = true)
                 }
 
+                override fun onInvalidItems(itemIds: List<String>) {
+                    handler.post {
+                        onInvalidItemsDetectedListener?.let { it(itemIds) }
```
Comment (cmaier) [inline core/src/main/java/io/snabble/sdk/shoppingcart/ShoppingCartUpdater.kt:105]: `onInvalidItemsDetectedListener?.invoke(itemIds)` is easier to read!

File: core/src/main/java/io/snabble/sdk/shoppingcart/ShoppingCart.kt [hunk 3]
```diff
@@ -55,6 +58,10 @@ class ShoppingCart(
     init {
         updateTimestamp()
         updater = project?.let { ShoppingCartUpdater(it, this) }
+        updater?.onInvalidItemsDetectedListener = { invalidItems ->
+            val items = filterNotNull().filter { it.id in invalidItems }
+            onInvalidItemsDetectedListener?.let { it(items) }
```
Comment (cmaier) [inline core/src/main/java/io/snabble/sdk/shoppingcart/ShoppingCart.kt:63]: `onInvalidItemsDetectedListener?.invoke(items)` is easier to read!

**Review by cmaier — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/snabble/sdk/shoppingcart/ShoppingCartUpdater.kt (hunk 2 above)
Comment (Fabtron) [reply core/src/main/java/io/snabble/sdk/shoppingcart/ShoppingCartUpdater.kt]: true, changed :D

**Review by cmaier — APPROVED** [review]
